test/e2e/network/netpol/k8s_util.go

```diff
@@ -0,0 +1,373 @@
+func (k *Kubernetes) CreateOrUpdateService(service *v1.Service) (*v1.Service, error) {
+	ns := service.Namespace
+	name := service.Name
+
+	createdService, err := k.ClientSet.CoreV1().Services(ns).Create(context.TODO(), service, metav1.CreateOptions{})
+	if err == nil {
+		log.Infof("created service %s/%s", ns, name)
+		return createdService, nil
+	}
+
+	log.Debugf("unable to create service %s/%s, let's try updating it instead (error: %s)", ns, name, err)
+	createdService, err = k.ClientSet.CoreV1().Services(ns).Update(context.TODO(), service, metav1.UpdateOptions{})
+	if err != nil {
+		log.Debugf("unable to update service %s/%s: %s", ns, name, err)
+	}
+
+	return createdService, err
+}
+
+// CreateOrUpdateDeployment is a convenience function for idempotent setup of deployments
+func (k *Kubernetes) CreateOrUpdateDeployment(deployment *appsv1.Deployment) (*appsv1.Deployment, error) {
+	ns := deployment.Namespace
+	log.Infof("creating/updating deployment %s/%s", ns, deployment.Name)
+
+	createdDeployment, err := k.ClientSet.AppsV1().Deployments(ns).Create(context.TODO(), deployment, metav1.CreateOptions{})
+	if err == nil {
+		log.Infof("created deployment %s/%s", ns, createdDeployment.Name)
+		return createdDeployment, nil
+	}
+
+	log.Debugf("unable to create deployment %s/%s, let's try update instead", ns, deployment.Name)
+	createdDeployment, err = k.ClientSet.AppsV1().Deployments(ns).Update(context.TODO(), createdDeployment, metav1.UpdateOptions{})
+	if err != nil {
+		log.Errorf("unable to create/update deployment %s/%s: %+v", ns, deployment.Name, err)
+	}
+
+	return createdDeployment, errors.Wrapf(err, "unable to update deployment %s/%s", ns, deployment.Name)
+}
+
+// CleanNetworkPolicies is a convenience function for deleting network policies before startup of any new test.
+func (k *Kubernetes) CleanNetworkPolicies(namespaces []string) error {
+	for _, ns := range namespaces {
+		log.Infof("deleting policies..........%v ", ns)
+		l, err := k.ClientSet.NetworkingV1().NetworkPolicies(ns).List(context.TODO(), metav1.ListOptions{})
+		if err != nil {
+			return errors.Wrapf(err, "unable to list network policies in ns %s", ns)
+		}
+		for _, np := range l.Items {
+			log.Infof("deleting network policy %s in ns %s", np.Name, ns)
+			err = k.ClientSet.NetworkingV1().NetworkPolicies(ns).Delete(context.TODO(), np.Name, metav1.DeleteOptions{})
+			if err != nil {
+				return errors.Wrapf(err, "unable to delete network policy %s", np.Name)
+			}
+		}
+	}
+	return nil
+}
+
+// ClearCache clears the kube pod cache
+func (k *Kubernetes) ClearCache() {
+	log.Info("Clearing pod cache")
+	k.mutex.Lock()
+	k.podCache = map[string][]v1.Pod{}
+	k.mutex.Unlock()
+	log.Info("Pod cache successfully cleared")
+}
+
+// CreateOrUpdateNetworkPolicy is a convenience function for updating/creating netpols
+func (k *Kubernetes) CreateOrUpdateNetworkPolicy(ns string, netpol *networkingv1.NetworkPolicy) (*networkingv1.NetworkPolicy, error) {
```

Comment:
We are testing NetworkPolicies so we should assert on all the operations related, if the test is supposed to create or update a network policy we should verify the operation is done correctly.
Also, this make tests hard to debug on failures, because we log 2 error messages

Reply:
yeah we can split this out

Reply:
👍